README.md
# meli-auth-sdk
[![Build Status](https://travis-ci.org/janis-commerce/meli-auth-sdk.svg?branch=master)](https://travis-ci.org/janis-commerce/meli-auth-sdk) [![Coverage Status](https://coveralls.io/repos/github/janis-commerce/meli-auth-sdk/badge.svg?branch=master)](https://coveralls.io/github/janis-commerce/meli-auth-sdk?branch=master)

Utilities for a simpler way to integrate serverless microservices with meli-auth service

## Installation

```sh
npm install @janiscommerce/meli-auth-sdk
```

## API
#### Methods
-   _async_  getAccessTokent(clientName, sellerId)
	|  Option     |  Type  |  Description  |
	|  ---------- | ------ | ------------- |
	|  clientName | string | Janis client name |
	|  sellerId   | string | Seller/User ID of MercadoLibre |
	|  **returns**| MeliToken | MercadoLibre authorization object |

#### Types
-   MeliToken
	|  Property    |  Type  |  Description  |
	|  ----------- | ------ | ------------- |
	|  accessToken | string | Mercadolibre access token |
	|  expiresIn   | string | Token expiration date in ISO 8601 format |

## Usage

```js
const MeliAuthSdk = require('@janiscommerce/meli-auth-sdk');
```

## Examples
```js
'use strict';

const { API } = require('@janiscommerce/api');
const MeliAuthSdk = require('@janiscommerce/meli-auth-sdk');
const requestPromise = require('request-promise');

class ExampleApi extends API {

	async process() {
		const orderId = '123455'
		const { accessToken, expiresIn } = await MeliAuthSdk.getAccessToken('test', '1234554');
		const response = await requestPromise({
			method: 'POST',
			uri: `https://api.mercadolibre.com/orders/${orderId}?access_token=${accessToken}`
			json: true
		});
		this.setBody({ response });
	}
}

module.exports = ExampleApi;
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2019-11-04
### Changed
- MS Call version updated to 3.0.0

## [1.0.0] - 2019-11-04
### Added
- Meli Auth SDK initial version


tests/meli-auth-sdk-test.js
'use strict';

const Settings = require('@janiscommerce/settings');
const MicroServiceCall = require('@janiscommerce/microservice-call');
const KmsEncryption = require('@janiscommerce/kms-encryption');
const assert = require('assert');
const sandbox = require('sinon').createSandbox();
const MeliAuthSdk = require('./../lib/meli-auth-sdk');
const MeliAuthSdkError = require('./../lib/meli-auth-sdk-error');

describe('MeliAuthSdk', () => {

	beforeEach(() => {
		this.keyArn = 'arn:aws:kms:us-east-1:026813942644:key/XXXXXXXX-XXXX-XXXX-XXXX-123456789876';
		this.stubSettingsGet = sandbox.stub(Settings, 'get').returns(this.keyArn);
		this.stubMscallGet = sandbox.stub(MicroServiceCall.prototype, 'get');
		this.stubKms = sandbox.stub(KmsEncryption.prototype, 'decrypt');
		this.getAccessToken = async () => MeliAuthSdk.getAccessToken('test', '12345');
	});

	afterEach(() => {
		sandbox.restore();
	});

	it('Should throw error for invalid clientName', () => {
		assert.rejects(MeliAuthSdk.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.INVALID_CLIENT_NAME
		});
	});

	it('Should throw error for invalid sellerId', () => {
		assert.rejects(MeliAuthSdk.getAccessToken('test'), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.INVALID_SELLER_ID
		});
	});

	it('Should get keyArn from settings', () => {
		// eslint-disable-next-line no-underscore-dangle
		assert.deepStrictEqual(MeliAuthSdk._getKmsArn(), this.keyArn);
	});

	it('Should get an error when keyArn not exists in settings', () => {
		this.stubSettingsGet.returns(null);
		// eslint-disable-next-line no-underscore-dangle
		assert.throws(() => { MeliAuthSdk._getKmsArn(); }, {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.ARN_NOT_FOUND
		});
	});

	it('Should return a token string', async () => {
		const msResponse = {
			statusCode: 200,
			body: {
				credentials: 'validcredentialencrypted-xxxsefweijio',
				expiresIn: '01-01-2019'
			}
		};
		const kmsResponse = {
			accessToken: 'testresulttoken'
		};

		const response = {
			accessToken: kmsResponse.accessToken,
			expiresIn: msResponse.body.expiresIn
		};

		this.stubMscallGet.resolves(msResponse);
		this.stubKms.resolves(kmsResponse);

		assert.deepEqual(await this.getAccessToken(), response);
	});

	it('Should fail ms request when status code is 200 and no response body', async () => {
		const msResponse = {
			statusCode: 200
		};

		this.stubMscallGet.resolves(msResponse);
		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.REMOTE_REQUEST_FAIL
		});
	});

	it('Should fail ms request when status code not 200', async () => {
		const msResponse = {
			statusCode: 403,
			body: {}
		};

		this.stubMscallGet.resolves(msResponse);
		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.REMOTE_REQUEST_FAIL
		});
	});

	it('Should fail ms request when ms call throws an error', async () => {
		const error = new Error('Random error');
		this.stubMscallGet.rejects(error);

		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.REMOTE_REQUEST_FAIL
		});
	});

	it('Should fail for no credentials in response', async () => {
		const msResponse = {
			statusCode: 200,
			body: {}
		};
		this.stubMscallGet.resolves(msResponse);

		// eslint-disable-next-line no-underscore-dangle
		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.CREDENTIALS_NOT_FOUND
		});

		assert.rejects(MeliAuthSdk.getAccessToken('test', '123'), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.CREDENTIALS_NOT_FOUND,
			message: 'No credentials found for seller: 123'
		});
	});

	it('Should fail for no expiration date in response', async () => {
		const msResponse = {
			statusCode: 200,
			body: {
				credentials: 'testing-testing-testing'
			}
		};
		this.stubMscallGet.resolves(msResponse);

		// eslint-disable-next-line no-underscore-dangle
		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.MALFORMED_RESPONSE
		});
	});

	it('Should fail for no token in decrypted data', async () => {
		const msResponse = {
			statusCode: 200,
			body: {
				credentials: 'validcredentialencrypted-xxxsefweijio',
				expiresIn: '01-01-2019',
				seller: 1234566
			}
		};
		const kmsResponse = {
			something: ''
		};

		this.stubMscallGet.resolves(msResponse);
		this.stubKms.resolves(kmsResponse);

		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.TOKEN_NOT_FOUND
		});

		assert.rejects(MeliAuthSdk.getAccessToken('test', '1234566'), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.TOKEN_NOT_FOUND,
			message: 'No token found for seller: 1234566'
		});

		this.stubKms.resolves(null);
		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.TOKEN_NOT_FOUND
		});

		this.stubKms.rejects('Rejection test');
		assert.rejects(this.getAccessToken(), {
			name: 'MeliAuthSdkError',
			code: MeliAuthSdkError.codes.TOKEN_NOT_FOUND
		});
	});

	it('Should fail kms decrypt function', async () => {
		this.stubKms.rejects(new Error('Rejection test'));
		const kms = new KmsEncryption({ keyArn: this.keyArn });
		assert.rejects(kms.decrypt(), {
			name: 'Error',
			message: 'Rejection test'
		});
	});

});


lib/meli-auth-sdk.js
'use strict';

const MicroServiceCall = require('@janiscommerce/microservice-call');
const KmsEncryption = require('@janiscommerce/kms-encryption');
const Settings = require('@janiscommerce/settings');
const MeliAuthSdkError = require('./meli-auth-sdk-error');

const KMS_ARN = 'kmsArn';

class MeliAuthSdk {

	static get kmsArnField() {
		return KMS_ARN;
	}

	/**
	 * Gets the KMS ARN from config file
	 * @returns {string} The kms arn
	 */
	static _getKmsArn() {
		const keyArn = Settings.get(MeliAuthSdk.kmsArnField);
		if(!keyArn) {
			throw new MeliAuthSdkError(
				`Missing kms config setting '${MeliAuthSdk.kmsArnField}'`,
				MeliAuthSdkError.codes.ARN_NOT_FOUND
			);
		}
		return keyArn;
	}

	/**
	 * @typedef {Object} MeliAuthCredentials
	 * @property {string} accessToken - Mercadolibre access token
	 * @property {string} refreshToken - Token used to get new access token
	 */

	/**
	 * @typedef {Object} MeliAuthResponse
	 * @property {number} seller - Mercadolibre seller id
	 * @property {string} credentials - Kms encrypted data
	 * @property {expiresIn} credentials - Token expiration date in ISO 8601 format
	 */

	/**
	 * Decrypts credentials string returned by ms meli-auth
	 * @static
	 * @param {string} keyArn KMS key used to decrypt
	 * @param {MeliAuthResponse} meliAuthResponse Response from ms meli-auth
	 * @returns {MeliAuthCredentials} Decrypted Mercadolibre credentials
	 */
	static async _kmsDecrypt(keyArn, meliAuthResponse) {
		const kmsEncryption = new KmsEncryption({ keyArn });
		const decrypt = {};
		try {
			decrypt.result = await kmsEncryption.decrypt(meliAuthResponse.credentials);
		} catch(kmsException) {
			decrypt.error = kmsException.message;
		}

		if(decrypt.error || !decrypt.result || !decrypt.result.accessToken) {
			throw new MeliAuthSdkError(`No token found for seller: ${meliAuthResponse.seller.toString()}`,
				MeliAuthSdkError.codes.TOKEN_NOT_FOUND);
		}

		return decrypt.result;
	}

	/**
	 * Requester to ms meli-auth
	 * @static
	 * @param {string} clientName Janis client name
	 * @param {string} sellerId Seller id of Mercadolibre
	 * @returns {MeliAuthResponse} Decrypted Mercadolibre credentials
	 */
	static async _msMeliAuthCall(clientName, sellerId) {
		const headers = {
			'janis-client': clientName
		};

		const service = new MicroServiceCall();
		const msCall = {};
		try {
			msCall.response = await service.get('meli-auth', 'credential', 'get', null, headers, { seller: sellerId });
		} catch(msException) {
			msCall.error = msException.message;
		}

		if(msCall.error || !msCall.response.body || msCall.response.statusCode !== 200) {
			throw new MeliAuthSdkError('Remote request to Mercadolibre authorization service failed -> ' + msCall.error,
				MeliAuthSdkError.codes.REMOTE_REQUEST_FAIL);
		}
		if(!msCall.response.body.credentials) {
			throw new MeliAuthSdkError(`No credentials found for seller: ${sellerId}`,
				MeliAuthSdkError.codes.CREDENTIALS_NOT_FOUND);
		}

		if(!msCall.response.body.expiresIn) {
			throw new MeliAuthSdkError('Malformed response from Mercadolibre authorization service',
				MeliAuthSdkError.codes.MALFORMED_RESPONSE);
		}

		return msCall.response.body;
	}

	/**
	 * @typedef {Object} MeliToken
	 * @property {string} accessToken - Mercadolibre access token
	 * @property {string} expiresIn - Token expiration date in ISO 8601 format
	 */

	/**
	 * Resolves the calls to ms meli-auth and decrypt the response
	 * @static
	 * @param {string} clientName Janis client name
	 * @param {string} sellerId Seller id of Mercadolibre
	 * @returns {MeliToken} Mercadolibre access token
	 */
	static async getAccessToken(clientName, sellerId) {
		if(!clientName)
			throw new MeliAuthSdkError('Invalid clientName', MeliAuthSdkError.codes.INVALID_CLIENT_NAME);

		if(!sellerId)
			throw new MeliAuthSdkError('Invalid sellerId', MeliAuthSdkError.codes.INVALID_SELLER_ID);

		const keyArn = this._getKmsArn();

		const msCall = await this._msMeliAuthCall(clientName, sellerId);

		const decryptedCredentials = await this._kmsDecrypt(keyArn, msCall);

		return {
			accessToken: decryptedCredentials.accessToken,
			expiresIn: msCall.expiresIn
		};
	}
}

module.exports = MeliAuthSdk;
